syntax = "proto3";

package railway;

import "google/protobuf/empty.proto";

service RailwayController {
    rpc List(RailwayListRequest) returns (stream Railway) {}
    rpc ListNeighborhood(NeighborhoodListRequest) returns (stream Neighborhood) {}
    rpc ListNeighborhoodOp(NeighborhoodOpListRequest) returns (stream NeighborhoodOp) {}
    rpc Retrieve(RailwayRetrieveRequest) returns (Railway) {}
    rpc RetrieveNeighborhood(NeighborhoodRetrieveRequest) returns (Neighborhood) {}
    rpc RetrieveNeighborhoodOp(NeighborhoodOpRetrieveRequest) returns (NeighborhoodOp) {}
    rpc Delete(RailwayRetrieveRequest) returns (Railway) {}
    rpc DeleteNeighborhood(NeighborhoodOpRetrieveRequest) returns (Railway) {}
    rpc Create(Railway) returns (Railway) {}
    rpc CreateNeigborhood(Neighborhood) returns (Neighborhood) {}
    rpc OptimizeNeighborhood(OptimizeRequest) returns (Optimized) {}
}

message RailwayListRequest {
    bool is_cont = 1;
}

message NeighborhoodListRequest {
}

message NeighborhoodOpListRequest {
}

message RailwayRetrieveRequest {
    int64 id = 1;
}

message NeighborhoodRetrieveRequest {
    int64 id = 1; 
}

message NeighborhoodOpRetrieveRequest {
    int64 id = 1; 
}

message OptimizeRequest {
}

message Railway {
    int64 id = 1;
    int64 iid = 2;
    string point = 3;
    bool is_cont = 4;
}

message Neighborhood {
    int64 id = 1;
    int64 source_id = 2;
    int64 target_id = 3;
    float length = 4;
}

message NeighborhoodOp {
    int64 id = 1;
    int64 source_id = 2;
    int64 target_id = 3;
    float length = 4;
}

message Optimized {
    string result = 1;
}